### ISSUE TITLE
no error when `Final` value reassigned in loop

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#2689]
### Summary

```py
from typing import Final

ABC: Final

for i in range(2):
    ABC = i  # expect error
```

but i hope that declaring a final variable in a loop is not removed:
```py
for i in range(10):
  i_str: Final = str(i)  # i like that this is allowed, please don't remove it
```

**Comment by Hugo-Polloli** [comment]
Look like this is already fixed on main, by https://github.com/astral-sh/ruff/commit/c056ee60ef1179839984b8a5f1216179f72da834
I’m opening a small PR to add mdtest regression coverage so this behavior stays protected 👍

**Comment by sharkdp** [comment]
Before we assert on this behavior in a test, I'd first like to understand why this should be allowed?



> ```py
> for i in range(10):
>   i_str: Final = str(i)  # i like that this is allowed, please don't remove it
> ```


Both mypy and pyright report an error (A "Final" variable cannot be assigned within a loop).

The [typing spec](https://typing.python.org/en/latest/spec/qualifiers.html#semantics-and-examples) has this to say:

> Note that a type checker need not allow Final declarations inside loops since the runtime will see multiple assignments to the same variable in subsequent iterations.

**Comment by KotlinIsland** [comment]
i've never understood this phrase of the spec, why would it ever be an issue to have a final variable in a for loop

**Comment by Hugo-Polloli** [comment]
> Before we assert on this behavior in a test, I'd first like to understand why this should be allowed?
> 
> > for i in range(10):
> >   i_str: Final = str(i)  # i like that this is allowed, please don't remove it
> 
> Both mypy and pyright report an error (A "Final" variable cannot be assigned within a loop).
> 
> The [typing spec](https://typing.python.org/en/latest/spec/qualifiers.html#semantics-and-examples) has this to say:
> 
> > Note that a type checker need not allow Final declarations inside loops since the runtime will see multiple assignments to the same variable in subsequent iterations.

My bad, I did not check specs and wrongly assumed that assigning loop-local Finals was spec-compliant for some reason.
I can either add a TODO in the mdtest on the loop-local Final assignment, or even better, I'd be happy to work on fixing that behavior, both are fine with me :)

**Comment by Hugo-Polloli** [comment]
> i've never understood this phrase of the spec, why would it ever be an issue to have a final variable in a for loop

I initially missed this too wanting to go too fast 😅 

But actually it makes sense because in Python a `for` loop does not introduce a new scope. Any assignment inside the loop targets the same binding each iteration, and that name is still visible after the loop (which should be "proof" enough that the binding is declared in the surrounding scope).
```python
for i in range(10):
    i_str: Final = str(i)
```

This does not mean "declare a fresh `Final` per iteration".
It’s more like "declare `i_str` as `Final` in the surrounding scope then assign to it repeatedly", ie. in effect:

```python
i_str: Final[str]
i_str = "0"
i_str = "1"   # second assignment to same Final
```

**Comment by KotlinIsland** [comment]
yes, I understand that. but my question is: what problem would it ever cause to annotate a variable as final in a for loop?

**Comment by Hugo-Polloli** [comment]
> yes, I understand that. but my question is: what problem would it ever cause to annotate a variable as final in a for loop?

I mean, for sure Python runtime itself has no issue with this. After all, `Final` has no runtime enforcement
It's purely a static typing concern: `Final` is a promise to the type checker that a name is only bound once, and a loop breaks that promise. So this is about consistency in how we treat that `Final` contract

Since the spec leaves this up to the type checker, it ultimately is a policy decision in how strict we want to be

**Comment by carljm** [comment]
I think a key factor here is our policy (which differs from both mypy and pyright) to allow re-declaration of a variable with a new type in the same scope, in general (aka shadowing). Because of our support for shadowing, we currently (I think intentionally?) allow this:

```py
from typing import Final

x: Final = 1
x: Final = "foo"
```

Given that that is allowed, it would seem inconsistent to me to disallow a `Final` declaration in a loop. (I believe that this consistency is precisely the reason that our implementation currently allows `Final` declaration in a loop: there is no special handling of `Final` in our treatment of loopback edges in control flow.)

If we decide that in general, once a variable is declared `Final` it cannot be shadowed, then we would disallow both my example above, and `Final` declaration in a loop.

(My personal opinion is that we don't need to add this restriction, allowing shadowing with explicit declaration is OK.)

**Comment by KotlinIsland** [comment]
> Final is a promise to the type checker that a name is only bound once

but in the case of a loop, even if it is literally the same variable being rebound, what tangible effect does this have on the semantics of the program? you can't **ever** access the previous value of the variable, so it's effectively a new variable

**Comment by carljm** [comment]
> but in the case of a loop, even if it is literally the same variable being rebound, what tangible effect does this have on the semantics of the program? you can't **ever** access the previous value of the variable, so it's effectively a new variable

Yes, this is effectively a re-statement of why we allow

```py
x: Final = 1
x: Final = "foo"
```

and variable shadowing with re-declaration in general.

The argument against this (that might be made by other type checkers, perhaps partly motivated also by implementation complexity) is not an argument of runtime unsafety, but just of user expectations and code clarity. (Which is why the counter-argument is "my type checker shouldn't be enforcing that.")
